From cdd2ac5c3fa69ff05d318bbe9e46e2b84b990dca Mon Sep 17 00:00:00 2001
From: "wang, biao" <biao.wang@intel.com>
Date: Mon, 10 Nov 2014 18:05:39 +0800
Subject: [PATCH] fix race of cpu dai active management

BZ: 226556

there is race of cpu dai active management

for example:
t0 call soc_compr_free_fe->soc_pcm_close to decrease cpu_dai->active while
another thread t1 call wm8994_put_class_w->snd_soc_dai_link_event to
decrease the same var.

this will lead to hwwdt due to sst shim register access while shim is
powered off.

Change-Id: I4d0b5325724887da0c082d36338dd9acafde3db6
Signed-off-by: Di Zhang <di.zhang@intel.com>
Signed-off-by: wang, biao <biao.wang@intel.com>
---
 sound/soc/soc-pcm.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/sound/soc/soc-pcm.c b/sound/soc/soc-pcm.c
index 429925c..4695f87 100644
--- a/sound/soc/soc-pcm.c
+++ b/sound/soc/soc-pcm.c
@@ -463,6 +463,7 @@ static int soc_pcm_close(struct snd_pcm_substream *substream)
 
 	mutex_lock_nested(&rtd->pcm_mutex, rtd->pcm_subclass);
 
+	mutex_lock_nested(&rtd->card->dapm_mutex, SND_SOC_DAPM_CLASS_RUNTIME);
 	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
 		cpu_dai->playback_active--;
 		codec_dai->playback_active--;
@@ -496,6 +497,7 @@ static int soc_pcm_close(struct snd_pcm_substream *substream)
 	if (platform->driver->ops && platform->driver->ops->close)
 		platform->driver->ops->close(substream);
 	cpu_dai->runtime = NULL;
+	mutex_unlock(&rtd->card->dapm_mutex);
 
 	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
 		if (!rtd->pmdown_time || codec->ignore_pmdown_time ||
@@ -1328,7 +1330,6 @@ int dpcm_be_dai_shutdown(struct snd_soc_pcm_runtime *fe, int stream)
 
 		dev_dbg(be->dev, "ASoC: close BE %s\n",
 			dpcm->fe->dai_link->name);
-
 		soc_pcm_close(be_substream);
 		be_substream->runtime = NULL;
 
-- 
1.9.1

